server/api/src/rsDataObjCopy.cpp

```diff
@@ -152,15 +161,49 @@ namespace {
         return destL1descInx;
     } // open_destination_data_obj
 
-    auto close_destination_data_obj(RsComm* rsComm, const int _inx) -> int
+    auto close_destination_data_obj(RsComm* _comm, const int _l1_index, const DataObjInp& _inp) -> int
     {
-        openedDataObjInp_t dataObjCloseInp{};
-        dataObjCloseInp.l1descInx = _inx;
-        dataObjCloseInp.bytesWritten = L1desc[L1desc[_inx].srcL1descInx].dataObjInfo->dataSize;
+        const std::string_view path = static_cast<const char*>(_inp.objPath);
+
+        log_api::trace("[{}:{}] - closing [{}]", __func__, __LINE__, path);
+
+        openedDataObjInp_t close_inp{};
+        close_inp.l1descInx = _l1_index;
 
-        rodsLog(LOG_DEBUG8, "[%s:%d] - closing [%s]", __FUNCTION__, __LINE__, L1desc[_inx].dataObjInp->objPath);
+        // Declare this here because we need to return an error if the source index is invalid.
+        int ec = 0;
+
+        // The L1 descriptor index is const and confirmed to be in-bounds above. Therefore, ignore linter.
+        //
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-constant-array-index)
+        const int source_l1_index = L1desc[_l1_index].srcL1descInx;
+        if (source_l1_index < 3 || source_l1_index >= NUM_L1_DESC) {
+            log_api::error("[{}:{}] - Source L1 descriptor for [{}] is out of range: [{}]",
+                           __func__,
+                           __LINE__,
+                           path,
+                           source_l1_index);
+
+            // Set the error code, but continue to the close call so the destination data object can be finalized.
+            ec = SYS_FILE_DESC_OUT_OF_RANGE;
+        }
+        else {
+            // The L1 descriptor index is const and confirmed to be in-bounds above. Therefore, ignore linter.
+            //
+            // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-constant-array-index)
+            close_inp.bytesWritten = L1desc[source_l1_index].dataObjInfo->dataSize;
+        }
+
+        if (const int close_ec = rsDataObjClose(_comm, &close_inp); ec >= 0) {
+            ec = close_ec;
+        }
```

Comment:
Wasn't sure whether the error code coming back from `rsDataObjClose` is more important than the source L1 descriptor being out of range.

Actually, looking at this again, we should probably set `L1desc[_l1_index].oprStatus` to `SYS_FILE_DESC_OUT_OF_RANGE` above so that close doesn't mark this replica as good.

Reply:
Does anything care about `.oprStatus`?

Reply:
`rsDataObjClose` uses `oprStatus` to determine whether to execute static post-PEPs and how to finalize data objects (i.e. if it's < 0, it can be marked stale). Seeing as `rsDataObjClose` is being used to close the two data objects in this API, that's why I feel it might be the right thing to do. Thoughts?

Reply:
Sounds reasonable.

Reply:
Done